### PR TITLE
[feature][ratelimiting] Allowing multiple rate limiting options per API

### LINKED ISSUE #205: [request] Multiple ratelimiting options per API
Here is our usecase: 

For an API, we impose a monthly limit of 100,000 calls. Also there is a second level throttling of 5 requests/second to avoid too many concurrent requests. Currently, this is not possible with kong. 

I think we can allow to have multiple rate limits per API provided they have different period values. And while throttling the requests, we can verify all of them and throw a 429 even if any one has crossed the quota. I would like a send a PR to fix this. But, before I would like to have the feedback. 

Thanks!

### PR DESCRIPTION
**Pull request by tamizhgeek**
This fixes issue #205 Please review and merge. 

Also this brings in a major change for the ratelimiting plugin configuration. Is there a way I can send  a PR for changing the docs as well?

### CODE REVIEW

File: kong/plugins/ratelimiting/access.lua [hunk 1]
```diff
@@ -13,29 +14,50 @@ function _M.execute(conf)
...
-  local remaining = conf.limit - current_usage
-  ngx.header[constants.HEADERS.RATELIMIT_LIMIT] = conf.limit
-  ngx.header[constants.HEADERS.RATELIMIT_REMAINING] = math.max(0, remaining - 1) -- -1 for this current request
+  for period, limit in pairs(periods_and_limits) do
+    limit = tonumber(limit)
+    -- Load current metric for configured period
+    local current_metric, err = dao.ratelimiting_metrics:find_one(ngx.ctx.api.id, identifier, current_timestamp, period)
+    if err then
+      ngx.log(ngx.ERR, tostring(err))
+      utils.show_error(500)
```
Comment (thibaultCha) [inline kong/plugins/ratelimiting/access.lua]: This is not available anymore, please rebase and use the new methods in #221.

I am also really not a fan of performing a SELECT in a loop. This could maybe be done in a simple SELECT call.

File: kong/plugins/ratelimiting/access.lua [hunk 2]
```diff
@@ -13,29 +14,50 @@ function _M.execute(conf)
...
+    local current_usage
+    if current_metric ~= nil then
+      current_usage = current_metric.value
+    else
+      current_usage = 0
+    end
+    
+    local remaining = limit - current_usage
+    -- Figure out the period, which has the least remaining calls left
+    if(remaining < least_remaining_limit["remaining"]) then
```
Comment (thibaultCha) [inline kong/plugins/ratelimiting/access.lua]: Our code style doesn't use parenthesis here

File: kong/plugins/ratelimiting/access.lua [hunk 3]
```diff
@@ -13,29 +14,50 @@ function _M.execute(conf)
...
 
-  -- What is the current usage for the configured period?
-  local current_usage
-  if current_metric ~= nil then
-    current_usage = current_metric.value
-  else
-    current_usage = 0
-  end
+  local least_remaining_limit = {}
+  least_remaining_limit["remaining"] = 999999999 -- A huge number so that any limit will be actually less than this
```
Comment (thibaultCha) [inline kong/plugins/ratelimiting/access.lua]: Please use a dot notation for those.

File: kong/plugins/ratelimiting/schema.lua [hunk 4]
```diff
@@ -1,6 +1,18 @@
 local constants = require "kong.constants"
+local stringy = require "stringy"
+local utils = require "kong.tools.utils"
+
+
+local function check_limit_period(value, table_t)
+   for _,v in ipairs(value) do
+      local parts = stringy.split(v, ':')
+      if utils.array_contains(constants.RATELIMIT.PERIODS, parts[1]) ~= true then
```
Comment (thibaultCha) [inline kong/plugins/ratelimiting/schema.lua]: Keeping this simple:

```lua
if not utils.array_contains(constants.RATELIMIT.PERIODS, parts[1]) then
```

**Comment by thibaultCha** [comment]
I wonder if we should show more infos in the headers now, such as:

```
X-RateLimit-Limit: hour:1000,second:10
X-RateLimit-Remaining: hour:990,second:0
``

File: kong/plugins/ratelimiting/access.lua [hunk 5]
```diff
@@ -13,29 +14,50 @@ function _M.execute(conf)
...
     identifier = ngx.var.remote_addr
   end
-
-  -- Load current metric for configured period
-  local current_metric, err = dao.ratelimiting_metrics:find_one(ngx.ctx.api.id, identifier, current_timestamp, conf.period)
-  if err then
-    ngx.log(ngx.ERR, tostring(err))
-    utils.show_error(500)
+  local periods_and_limits = {}
+  if conf.limit then
```
Comment (thibaultCha) [inline kong/plugins/ratelimiting/access.lua]: `limit` is a required property, it shouldn't be necessary to test if it is set.

**Comment by thibaultCha** [comment]
Finally, I also believe this can do the same thing (find the least remaining limit) in fewer lines of code and only 1 DB call (that needs to be implemented):

```lua
local least_remaining_limit = {}

local current_metrics, err = dao.ratelimiting_metrics:find_for_period(ngx.ctx.api.id, identifier, current_timestamp) -- could be implemented
if err then
  -- ...
end

for _, period_conf in ipairs(conf.limit) do
  local period, period_limit = unpack(stringy.split(period_conf, ":"))
  period_limit = tonumber(period_limit)
  period_metric = current_metrics[period]

  -- What is the current usage for this period?
  local period_usage = period_metric and period_metric.value or 0
  local period_remaining = period_limit - period_usage
  
  -- Figure out the period which has the least remaining calls left
  if not least_remaining_limit.remaining or period_remaining < least_remaining_limit.remaining then
    least_remaining_limit.period = period
    least_remaining_limit.period_limit = period_limit
    least_remaining_limit.remaining = period_remaining
  end
end
```

It also uses more descriptive variable names.

But because of my comments about the Headers, I am not sure this is what we really want. We should probably inform the user of all his limits, and just test if at least 1 of those rules is over the current limit.

**Comment by montanaflynn** [comment]
I agree that we should show all the limits on an API, I prefer separate headers like this: 

```
X-RateLimit-Limit-Hour: 1000
X-RateLimit-Limit-Second: 10
X-RateLimit-Remaining-Hour: 990
X-RateLimit-Remaining-Second: 0
```
Additionally this PR seems to be missing the `X-Rate-Limit-Reset` header which is standard and returns the time at which the current rate limit window resets in UTC epoch seconds.

File: kong/plugins/ratelimiting/access.lua (hunk 1 above)
Comment (tamizhgeek) [inline kong/plugins/ratelimiting/access.lua]: @thibaultCha I don't think so, considering how the partition key is defined for ratelimiting_metrics. 

```
PRIMARY KEY ((api_id, identifier, period_date, period))
```
I can't query the table without specifying a equals operator for 'period' column. Any other help is appreciated. I'm not that great in doing cassandra data modelling.

**Comment by tamizhgeek** [comment]
@thibaultCha all feedback comments incorporated, except the multiple SELECT calls.

**Comment by montanaflynn** [comment]
@tamizhgeek I was able to find [an API](http://api.ferret-go.com/docs/account.html) returning headers for multiple ratelimit timeframes, could you make the headers be in this format? `X-RateLimit-Day-Limit: 10000`

@thibaultCha we can create a new issue for the `X-RateLimit-*-Reset` header if you want to merge this.

**Comment by tamizhgeek** [comment]
@montanaflynn Will do. The Day-Limit makes more sense than Limit-Day.

File: kong/plugins/ratelimiting/access.lua [hunk 6]
```diff
@@ -14,23 +15,44 @@ function _M.execute(conf)
...
     identifier = ngx.var.remote_addr
   end
+  
 
-  -- Load current metric for configured period
-  local current_metric, err = dao.ratelimiting_metrics:find_one(ngx.ctx.api.id, identifier, current_timestamp, conf.period)
-  if err then
-    return responses.send_HTTP_INTERNAL_SERVER_ERROR(err)
-  end
+  local least_remaining_limit = 99999999 -- A huge number so that any limit will be actually less than this
```
Comment (thibaultCha) [inline kong/plugins/ratelimiting/access.lua]: This is still very wrong, please see the code-snippet I posted previously.

File: kong/plugins/ratelimiting/access.lua [hunk 7]
```diff
@@ -14,23 +15,44 @@ function _M.execute(conf)
...
+      return responses.send_HTTP_INTERNAL_SERVER_ERROR(err)
+    end
+
+    -- What is the usage for the configured period?
+    local period_usage
+    if period_metric ~= nil then
+      period_usage = period_metric.value
+    else
+      period_usage = 0
+    end
```
Comment (thibaultCha) [inline kong/plugins/ratelimiting/access.lua]: This was also updated

File: kong/plugins/ratelimiting/access.lua [hunk 8]
```diff
@@ -14,23 +15,44 @@ function _M.execute(conf)
...
-  local current_metric, err = dao.ratelimiting_metrics:find_one(ngx.ctx.api.id, identifier, current_timestamp, conf.period)
-  if err then
-    return responses.send_HTTP_INTERNAL_SERVER_ERROR(err)
-  end
+  local least_remaining_limit = 99999999 -- A huge number so that any limit will be actually less than this
+
+  for _, period_conf in ipairs(conf.limit) do
+    local period, limit = unpack(stringy.split(period_conf, ":"))
+    
+    limit = tonumber(limit)
```
Comment (thibaultCha) [inline kong/plugins/ratelimiting/access.lua]: `period_limit`, like in the code snippet. I find this to be very confusing and would like other contributors to quickly understand this code.

File: kong/plugins/ratelimiting/access.lua [hunk 9]
```diff
@@ -14,23 +15,44 @@ function _M.execute(conf)
...
+
+    -- What is the usage for the configured period?
+    local period_usage
+    if period_metric ~= nil then
+      period_usage = period_metric.value
+    else
+      period_usage = 0
+    end
+    
+    local remaining = limit - period_usage
```
Comment (thibaultCha) [inline kong/plugins/ratelimiting/access.lua]: `period_remaining`

File: kong/plugins/ratelimiting/access.lua [hunk 10]
```diff
@@ -14,23 +15,44 @@ function _M.execute(conf)
...
-  ngx.header[constants.HEADERS.RATELIMIT_REMAINING] = math.max(0, remaining - 1) -- -1 for this current request
+    -- Figure out the period, which has the least remaining calls left
+    if remaining < least_remaining_limit then
+       least_remaining_limit = remaining
+    end
 
-  if remaining == 0 then
+    -- Set the reamining limit of this period in the header
+    ngx.header[constants.HEADERS.RATELIMIT_LIMIT.."-"..string.upper(period)] = limit
+    ngx.header[constants.HEADERS.RATELIMIT_REMAINING.."-"..string.upper(period)] = math.max(0, remaining - 1) -- -1 for this period request
```
Comment (thibaultCha) [inline kong/plugins/ratelimiting/access.lua]: :+1:

**Comment by thibaultCha** [comment]
Could you sqash your commits please? We like to keep a clean git history. I will try to come up with a SELECT request for that use case later, thanks!

File: kong/tools/faker.lua [hunk 11]
```diff
@@ -33,16 +33,27 @@ Faker.FIXTURES = {
...
     -- DEVELOPMENT APIs. Please do not use those in tests
     { name = "API DEV 1", public_dns = "dev.com", target_url = "http://mockbin.com" },
   },
   consumer = {
     { custom_id = "provider_123" },
-    { custom_id = "provider_124" }
+    { custom_id = "provider_124" },
+    { custom_id = "provider_125" },
   },
+  keyauth_credential = {
```
Comment (thibaultCha) [inline kong/tools/faker.lua]: Why is this here? They are supposed to be under `plugins_configuration` and the comment "TODO: remove plugins from core" is **not** about `plugins_configuration`, but about the `keyauth_credentials` and `basicauth_credentials`.

File: kong/plugins/ratelimiting/access.lua [hunk 12]
```diff
@@ -14,23 +15,39 @@ function _M.execute(conf)
...
+    local period_remaining = period_limit - period_usage
+
+    -- Figure out the period, which has the least remaining calls left
+    if not least_remaining_limit or period_remaining < least_remaining_limit then
+       least_remaining_limit = period_remaining
+    end
+
+    -- Set the reamining limit of this period in the header
+    ngx.header[constants.HEADERS.RATELIMIT_LIMIT:gsub("<duration>", string.upper(period))] = period_limit
+    ngx.header[constants.HEADERS.RATELIMIT_REMAINING:gsub("<duration>", string.upper(period))] = math.max(0, period_remaining - 1) -- -1 for this period request
```
Comment (thibaultCha) [inline kong/plugins/ratelimiting/access.lua]: Isn't this going to be `X-RateLimit-SECOND-Remaining` ?

**Comment by thibaultCha** [comment]
This is the big issue with Cassandra: the way you will query your data influences the way you design your schema. Now that our usage of the ratelimiting table changed, we should review the schema because it is not efficient anymore. And that means big troubles: migrations of data etc. Not even mentioning we still don't remove the metrics once they become obsolete and that is going to take a **lot** of space eventually. We either need to design the plugin and the model to do that, or write a cleanup script to execute at regular interval.

Anyways. Could you please squash your commits into 1 or 2 please @tamizhgeek ? I will do the cleanup after merging.

File: kong/plugins/ratelimiting/access.lua (hunk 12 above)
Comment (tamizhgeek) [inline kong/plugins/ratelimiting/access.lua]: I thought the headers are case-insensitive anyway!

File: kong/tools/faker.lua (hunk 11 above)
Comment (tamizhgeek) [inline kong/tools/faker.lua]: Oops misunderstood. Will replace the TODO

**Comment by tamizhgeek** [comment]
@thibaultCha I was frequently rebasing with the mashape/kong master all along the way. Now if I squash the commits, it's gonna rewrite the history and git will behave very weirdly. Also it's always not a good idea to rewrite commits when somebody else has someone else's commit also, right?

**Comment by tamizhgeek** [comment]
@thibaultCha Latest comment feedback incorporated and also squashed into two commits. Please let me know if there is some screw up in the git history

**Comment by thibaultCha** [comment]
Something went clearly wrong with your squashing. We should only see 2 commits in this PR, instead I am seeing 42. Please fix it before merging.

I would suggest a quick solution:

```
git checkout master # make sure you're on master
git branch old-master # copy the branch
git fetch upstream # fetch our current master branch
git reset --hard upstream/master # reset it with our current history
git cherry-pick 692bae2e8b082de3e562849dd6af56f850abe048
git cherry-pick 99c383db87783acc88318aea2dd10641099a529a
git push --force
# if all went well
git branch -D old-master
```

**Comment by tamizhgeek** [comment]
@thibaultCha That worked well! Thanks.

**Comment by tamizhgeek** [comment]
@thibaultCha Looks like there is again some changes in the way test fixtures are created. Should I wait some more time before any other conflicting changes come in? Or can I go ahead, resolve conflicts and update the PR. Is there anything else blocking you from merging this?

**Comment by thibaultCha** [comment]
Yes there are two things preventing us from merging this:

- The conflicts, I see you resolved them so that is great
- The fact that the performance of this plugin now depends on how many limits one sets. I do think this change improves the usability, I like the `hour:X` notation. But as discussed, the more limits you add to an API, the slower it is going to be. Say someone adds 5 periods-limit on an API, it is now making 5 queries to Cassandra for every request :/. It is something we discussed with @thefosk and we are not sure about what to do. We do not know what would be the consequences on high usage, maybe this could be benchmarked. One way to partially improve this would be to break from the loop once a limit has been broken, but then the user won't have all the `X-Ratelimit-%s-Remaining` headers. Another way would be to rewrite the ratelimiting table and find a way to increment and retrieval in one request each, which might not even be possible.

So that's it, so far nobody stepped in about this discussion... Personally, I will add as a note that this change doesn't make it less performant for configurations equal to the old plugin (1 period = 1 query), but people might want to add 3, 4, 5 periods and quickly end up with some serious DB querying.

**Comment by tamizhgeek** [comment]
@thibaultCha I totally agree with your views. Probably I should do a benchmark with multiple limits set and see how well it performs. Other than that, I'm still not very clear on the status of this PR. Should you include more people who can help us make this decision? I'm pressing this because this is one of the show stoppers for us to adopt kong. If this PR isn't gonna make it in - the only option left for us to write a custom plugin mimicking this functionality and keep it out of core. 

Anyway even to do that, we need some benchmarks on our side. I'll do the benchmarking and publish the results. But it'd be good to know whether this PR would make the cut and be available for 0.3 release, atleast with the note you mentioned.

**Comment by sinzone** [comment]
@tamizhgeek we def need a benchmark just to get an idea on how well this is going to work before merging. 

Other than rewrite the ratelimiting table and find a way to increment and retrieval in one request each as @thibaultCha mentioned, consider that 90% of the API creators will put only 2 limit max. Looking at our mashape marketplace most have 1 or 2 limits (for free APIs).

Another thing to consider: Billing plugin. If rate limit will be the foundation to set up pricing plans and their limits, then an avg API has 3 plans with 2 limits in each plan. So that is 6 limits x API.

**Comment by thibaultCha** [comment]
@tamizhgeek This PR can totally make it if the benchmarks don't go exponentially over the roof, so don't worry.

Like @sinzone said, which I partially mentioned by saying this doesn't make it less performant than previously (because previously we only had 1 limit anyways), providers probably won't have too many requests. But building software with this kind of assumption is bad. A simple example: we're soon gonna add an `overall:X` limit, which might be one more query. Anyways, I don't think the benchmarks will go over the roof either, so unless it does, we'll be really happy about this (like I said, I also like the new notation for the limits, which is already a cool improvement).

**Comment by sinzone** [comment]
should we create a new issue for overall:X?

**Comment by thibaultCha** [comment]
Yeah let's do it

**Comment by sinzone** [comment]
@tamizhgeek you have some benchmarks? We're getting close to the release date and we want to be sure that this PR is going to make it.

**Comment by tamizhgeek** [comment]
I don't have yet. I am on a vacation for the past week. Will get it done on
Monday once i am back at work.

**Comment by sinzone** [comment]
@tamizhgeek we're few days away from the 0.3. we need something within the next 48hours.

**Comment by Jazzatola** [comment]
+1 for multiple rate limits.

**Comment by thefosk** [comment]
I actually prefer more having the following format `value.second=100&value.hour=200`, it will make the configuration and the schema more explicit, without worrying about parsing a custom format. Will submit a PR soon.

**Comment by thefosk** [comment]
Closed with #382.
